feat: Decouple properties in timeline animations & support parallel animation

TremblingMoeNew opened the pull request:
#807
# 介绍
对时间线动画中的变换参数进行解耦，不再捆绑更新所有参数。并添加了setTransform、setTempAnimation、setAnimation指令的平行变换动画支持，允许对同一个目标同时进行多条针对不同变换参数的变换动画指令
# 更改
1. Perform执行完成时的unmount动作改为直接使用perform对象、而非使用名称，从而修正了原先连续执行设置了-continue参数的setTransform指令时无法正确运行的问题
2. 修改时间线动画相关系统，不再强制捆绑更新全部参数
3. 为setTransform、setTempAnimation、setAnimation指令添加了-parallel参数。添加该参数时，该指令不会打断对同一target正在执行的其他变换动画指令，从而进行互不干扰的并行运行。
# 注意事项
1. 用户需自行保证平行执行的变换没有对同样的属性进行修改
2. 用户需自行注意并规避指令执行被nextSentence()调用打断造成演出中断这一旧有问题的影响

TremblingMoeNew committed:
Fix: setTransform cannot execute continuously when targeting the same object

TremblingMoeNew committed:
Feat: Decouple properties in timeline animations & support parallel animation for setTransform command

TremblingMoeNew committed:
Feat: Support parallel animation for setTempAnimation command

TremblingMoeNew committed:
Feat: Support parallel animation for setAnimation command; Fix: potential runtime error caused by undefined parameters reported by review bot

TremblingMoeNew commented:
添加了setAnimation的-parallel支持，并添加了一个演示平行变换动画使用的demo剧本文件

HardyNLee commented on packages/webgal/src/Core/gameScripts/setTransform.ts:
```diff
@@ -43,8 +45,7 @@ export const setTransform = (sentence: ISentence): IPerform => {
   WebGAL.animationManager.addAnimation(newAnimation);
-  const animationDuration = getAnimateDuration(animationName);
-
+  const animationDuration = getAnimateDurationFromObj(newAnimation);
```
实际上原来的函数就可以正常使用，可以看一下上一句的
``` ts
WebGAL.animationManager.addAnimation(newAnimation);
```
和 `getAnimateDuration` 的实现

setTempAnimation 那边的改动也是同理

HardyNLee reviewed (changes requested):
虽然感觉为了规避 `unmountPerform(performName: string)` 而新加入 `isParallel` 有点把演出系统搞复杂了，但从功能上看确实成功的实现了平行动画。虽然把 `ITransform` 的所有属性变可选看着有点冒险，但因为出场时肯定已经融合了一遍完整的 `baseTransform`，所以存档也没什么问题，就是可能以后要注意一下 baseTransform 得写全。

其中最需要修改的应该是 lint 和构建时 tsc 检查不过需要更改，其他可改可不改吧。

TremblingMoeNew committed:
Merge remote-tracking branch 'origin/dev' into feat/parallel-animation-support

TremblingMoeNew committed:
Merge remote-tracking branch 'origin/dev' into feat/parallel-animation-support

HardyNLee reviewed (approved):
# 简介

下文
- 简称 `setAnimation` `setTransform` `setTempAnimation` 为 __Timeline 动画演出__

---

# 改造演出系统

此 PR 为 Timeline 动画演出命令添加了 `-parallel` 参数, 实现了对同目标同时使用两个 timeline 动画, 同时对演出系统做了些改造;

-parallel 与其说是平行动画, 原理上不如说是"不打断上一个相同目标的动画", 几乎所有的改造都是围绕这点进行;

这是因为在此之前, 如果有作用于同目标的多个动画, 前面的动画必须被强制完成, 然后开始最靠后的动画, 此 PR 从几个方面着手:

1. 有 `-parallel` 参数时, 不使用 `unmountPerform` 打断上一个同目标的动画
2. IPerform 倒计时结束, 不使用 `unmountPerform(performName: string)` 结束演出
  (1). 因为这有可能错误地把另一个同目标的动画演出结束, 所有的动画类演出 `performName` 都是 `animation-${target}`
  (2). 新造了一个函数 `softUnmountPerform(perform: IPerform)`, 倒计时结束直接结束自身
3. `IPerform` 新增了一个参数 `isParallel` 参数
  这是由于 `arrangeNewPerform` 会检查是否有同名演出, 有就干掉所有同名演出, 新增该参数是为了避开这个问题

理解这几点, 就容易理解绝大部分更改的出发点, 这么做似乎也没什么问题

# 改造 ITransform

剩下的更改是把 `ITransform` 所有参数都改成可选的了;

因为此 PR 之前在生成 timeline 对象前, 会先融合一遍现有的 transform, 出于 _一些原因_ , 现有的 transform 永远是完整的, 所以这会导致无法使用平行动画, 因为一个 timeline 动画会 set 所有的参数, 平行动画需要的是按需 set 参数

### 这是否会导致一些问题

从存档的角度出发应该没什么问题, 因为对象在出场时必先融合一遍完整的 transform, 所有存档应该永远是完整的;

唯一需要注意的是, 以后再往 ITransform 加属性, 一定要把 `baseTransform` 写完整, 现在参数可选了之后是不会有语法提示的;

---

实际测试了一段时间, 在编辑器那边也没有问题, 我觉得此 PR 可被合并

TremblingMoeNew committed:
Merge branch 'dev' into feat/parallel-animation-support

TremblingMoeNew committed:
Merge remote-tracking branch 'origin/dev' into feat/parallel-animation-support

MakinoharaShoko committed:
fix: parallel perform dedup & prefix matching for correct save/load

- Use unique performName (baseName#animationName) for parallel performs
- Add prefix matching in unmountPerform and removePerformByName
- Fix force path splice ordering (before goNextWhenOver)
- Use baseTransform as default in updateEffect else path
- Remove isParallel dedup guard (unique IDs handle it naturally)

MakinoharaShoko committed:
feat: add writeFullEffect parameter to getAnimationObject and update related functions
